File: adapter/content.py
from .jsonutil import json_dumps


def content_to_chat(content):
    if content is None:
        return ""
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts = []
        has_non_text = False
        for item in content:
            if isinstance(item, str):
                parts.append({"type": "text", "text": item})
                continue
            if not isinstance(item, dict):
                parts.append({"type": "text", "text": str(item)})
                continue
            item_type = item.get("type")
            if item_type in ("input_text", "output_text", "text"):
                parts.append({"type": "text", "text": str(item.get("text", ""))})
            elif item_type in ("input_image", "image_url"):
                image_url = item.get("image_url") or item.get("url")
                if not image_url and item.get("image_base64"):
                    mime = item.get("media_type") or "image/png"
                    image_url = f"data:{mime};base64,{item['image_base64']}"
                if image_url:
                    parts.append({"type": "image_url", "image_url": {"url": image_url}})
                    has_non_text = True
            elif "content" in item:
                nested = content_to_chat(item["content"])
                if isinstance(nested, list):
                    parts.extend(nested)
                    has_non_text = True
                elif nested:
                    parts.append({"type": "text", "text": nested})
            else:
                parts.append({"type": "text", "text": json_dumps(item)})
        if has_non_text:
            return parts
        return "\n".join(part["text"] for part in parts if part.get("text"))
    if isinstance(content, dict):
        if "text" in content:
            return str(content["text"])
        if "content" in content:
            return content_to_chat(content["content"])
    return str(content)


IMAGE_TYPE_MARKERS = {"input_image", "image_url"}
IMAGE_KEY_MARKERS = {"image_url", "image_base64"}


def value_has_image(value):
    if value is None:
        return False
    if isinstance(value, list):
        return any(value_has_image(item) for item in value)
    if isinstance(value, dict):
        item_type = value.get("type")
        if item_type in IMAGE_TYPE_MARKERS:
            return True
        if any(key in value and value.get(key) for key in IMAGE_KEY_MARKERS):
            return True
        return any(value_has_image(child) for child in value.values())
    return False


def request_has_image(request):
    return value_has_image(request)

File: internal/content/content.go
package content

import (
	"fmt"
)

func ToChat(value any) any {
	if value == nil {
		return ""
	}
	switch v := value.(type) {
	case string:
		return v
	case []any:
		parts := make([]map[string]any, 0, len(v))
		hasNonText := false
		for _, item := range v {
			switch it := item.(type) {
			case string:
				parts = append(parts, map[string]any{"type": "text", "text": it})
			case map[string]any:
				itemType, _ := it["type"].(string)
				switch itemType {
				case "input_text", "output_text", "text":
					parts = append(parts, map[string]any{"type": "text", "text": fmt.Sprint(it["text"])})
				case "input_image", "image_url":
					imageURL := imageURLValue(firstNonNil(it["image_url"], it["url"]))
					if imageURL == "" {
						if b64, ok := it["image_base64"].(string); ok && b64 != "" {
							mime, _ := it["media_type"].(string)
							if mime == "" {
								mime = "image/png"
							}
							imageURL = "data:" + mime + ";base64," + b64
						}
					}
					if imageURL != "" {
						parts = append(parts, map[string]any{"type": "image_url", "image_url": map[string]any{"url": imageURL}})
						hasNonText = true
					}
				default:
					if nested, ok := it["content"]; ok {
						converted := ToChat(nested)
						if nestedParts, ok := converted.([]any); ok {
							for _, nestedPart := range nestedParts {
								if part, ok := nestedPart.(map[string]any); ok {
									parts = append(parts, part)
								} else {
									parts = append(parts, map[string]any{"type": "text", "text": fmt.Sprint(nestedPart)})
								}
							}
							hasNonText = true
						} else if s, ok := converted.(string); ok && s != "" {
							parts = append(parts, map[string]any{"type": "text", "text": s})
						}
					} else {
						parts = append(parts, map[string]any{"type": "text", "text": fmt.Sprint(it)})
					}
				}
			default:
				parts = append(parts, map[string]any{"type": "text", "text": fmt.Sprint(item)})
			}
		}
		if hasNonText {
			out := make([]any, 0, len(parts))
			for _, part := range parts {
				out = append(out, part)
			}
			return out
		}
		text := ""
		for _, part := range parts {
			if s, ok := part["text"].(string); ok && s != "" {
				if text != "" {
					text += "\n"
				}
				text += s
			}
		}
		return text
	case map[string]any:
		if text, ok := v["text"]; ok {
			return fmt.Sprint(text)
		}
		if c, ok := v["content"]; ok {
			return ToChat(c)
		}
	}
	return fmt.Sprint(value)
}

func RequestHasImage(request map[string]any) bool {
	return ValueHasImage(request)
}

func ValueHasImage(value any) bool {
	switch v := value.(type) {
	case []any:
		for _, item := range v {
			if ValueHasImage(item) {
				return true
			}
		}
	case map[string]any:
		itemType, _ := v["type"].(string)
		if itemType == "input_image" || itemType == "image_url" {
			if imageURLValue(firstNonNil(v["image_url"], v["url"])) != "" {
				return true
			}
			if b64, ok := v["image_base64"].(string); ok && b64 != "" {
				return true
			}
		}
		for _, child := range v {
			if ValueHasImage(child) {
				return true
			}
		}
	}
	return false
}

func imageURLValue(value any) string {
	switch v := value.(type) {
	case string:
		return v
	case map[string]any:
		if url, ok := v["url"].(string); ok {
			return url
		}
	}
	return ""
}

func firstNonNil(values ...any) any {
	for _, value := range values {
		if value != nil {
			return value
		}
	}
	return nil
}
